packages/core/core/src/requests/ParcelConfigRequest.js

```diff
@@ -1,28 +1,84 @@
-// @flow
+// @flow strict-local
 import type {
   FilePath,
   RawParcelConfig,
   ResolvedParcelConfigFile,
   PackageName,
 } from '@parcel/types';
+import type {StaticRunOpts} from '../RequestTracker';
 import type {
+  ExtendableParcelConfigPipeline,
   ParcelOptions,
   ProcessedParcelConfig,
-  ExtendableParcelConfigPipeline,
-} from './types';
+} from '../types';
 
 import {resolveConfig, resolve, validateSchema} from '@parcel/utils';
 import ThrowableDiagnostic from '@parcel/diagnostic';
+// $FlowFixMe
 import {parse} from 'json5';
 import path from 'path';
 import assert from 'assert';
 
-import ParcelConfig from './ParcelConfig';
-import ParcelConfigSchema from './ParcelConfig.schema';
+import ParcelConfig from '../ParcelConfig';
+import ParcelConfigSchema from '../ParcelConfig.schema';
 
 type ConfigMap<K, V> = {[K]: V, ...};
 
-export default async function loadParcelConfig(options: ParcelOptions) {
+type ConfigAndRef = {|
+  config: ProcessedParcelConfig,
+  configRef: number,
+|};
+
+type RunOpts = {|
+  input: null,
+  ...StaticRunOpts,
+|};
+
+export type ParcelConfigRequest = {|
+  id: string,
+  type: string,
+  input: null,
+  run: () => Promise<ConfigAndRef>,
+|};
+
+const type = 'parcel_config_request';
+
+export default function createParcelConfigRequest() {
+  let disposeConfigRef: void | (() => Promise<mixed>);
+  return {
+    id: type,
+    type,
+    async run({api, options, farm}: RunOpts): Promise<ConfigAndRef> {
+      let {config, extendedFiles} = await loadParcelConfig(options);
+      let processedConfig = config.getConfig();
+      let {ref, dispose} = await farm.createSharedReference(processedConfig);
+      disposeConfigRef && (await disposeConfigRef());
+      disposeConfigRef = dispose;
+
+      api.invalidateOnFileUpdate(config.filePath);
+      api.invalidateOnFileDelete(config.filePath);
+
+      for (let filePath of extendedFiles) {
+        api.invalidateOnFileUpdate(filePath);
+        api.invalidateOnFileDelete(filePath);
+      }
+
+      if (config.filePath === options.defaultConfig?.filePath) {
+        api.invalidateOnFileCreate('**/.parcelrc');
+      }
+
+      // Need to do this because of reinstantiate the shared reference
```

Comment:
😞 could that happen on demand somehow?

Reply:
This is no longer an issue with the commit I just pushed. Basically I made the config a cache entry instead of a shared reference. For now I have some basic caching so that the workers don't read the cache entry over and over again (https://github.com/parcel-bundler/parcel/blob/request-graph-enhancements/packages/core/core/src/worker.js#L34-L35). Was thinking that wouldn't be necessary when we have an in memory cache layer. Technically the simple map cache could keep growing, but in practice I don't think that will be an issue since most people probably won't be changing their parcel config a bunch in watch mode. Let me know if you think I need to address that though.